agency/agency_impl.go

```diff
@@ -0,0 +1,394 @@
+	}
+	//fmt.Printf("Agent response: %s\n", string(raw))
+	elems, err := resp.ParseArrayBody()
+	if err != nil {
+		return driver.WithStack(err)
+	}
+	if len(elems) != 1 {
+		return driver.WithStack(fmt.Errorf("Expected 1 element, got %d", len(elems)))
+	}
+	// If empty key parse directly
+	if len(key) == 0 {
+		if err := elems[0].ParseBody("", &value); err != nil {
+			return driver.WithStack(err)
+		}
+	} else {
+		// Now remove all wrapping objects for each key element
+		var rawObject map[string]interface{}
+		if err := elems[0].ParseBody("", &rawObject); err != nil {
+			return driver.WithStack(err)
+		}
+		var rawMsg interface{}
+		for keyIndex := 0; keyIndex < len(key); keyIndex++ {
+			if keyIndex > 0 {
+				var ok bool
+				rawObject, ok = rawMsg.(map[string]interface{})
+				if !ok {
+					return driver.WithStack(fmt.Errorf("Data is not an object at key %s", key[:keyIndex+1]))
+				}
+			}
+			var found bool
+			rawMsg, found = rawObject[key[keyIndex]]
+			if !found {
+				return driver.WithStack(KeyNotFoundError{Key: key[:keyIndex+1]})
+			}
+		}
+		// Encode to json ...
+		encoded, err := json.Marshal(rawMsg)
+		if err != nil {
+			return driver.WithStack(err)
+		}
+		// and decode back into result
+		if err := json.Unmarshal(encoded, &value); err != nil {
+			return driver.WithStack(err)
+		}
+	}
+
+	//	fmt.Printf("result as JSON: %s\n", rawResult)
+	return nil
+}
+
+type writeUpdate struct {
+	Operation string      `json:"op,omitempty"`
+	New       interface{} `json:"new,omitempty"`
+	TTL       int64       `json:"ttl,omitempty"`
+	URL       string      `json:"url,omitempty"`
+}
+
+type writeCondition struct {
+	Old      interface{} `json:"old,omitempty"`      // Require old value to be equal to this
+	OldEmpty *bool       `json:"oldEmpty,omitempty"` // Require old value to be empty
+	IsArray  *bool       `json:"isArray,omitempty"`  // Require old value to be array
+}
+
+type writeTransaction []map[string]interface{}
+type writeTransactions []writeTransaction
+
+type writeResult struct {
+	Results []int64 `json:"results"`
+}
+
+// WriteKey writes the given value with the given key with a given TTL (unless TTL is zero).
+// If you pass a condition (only 1 allowed), this condition has to be true,
+// otherwise the write will fail with a ConditionFailed error.
+func (c *agency) WriteKey(ctx context.Context, key []string, value interface{}, ttl time.Duration, condition ...WriteCondition) error {
+	var cond WriteCondition
+	switch len(condition) {
+	case 0:
+	// No condition, do nothing
+	case 1:
+		cond = condition[0]
+	default:
+		return driver.WithStack(fmt.Errorf("too many conditions"))
+	}
+	if err := c.write(ctx, key, value, cond, ttl); err != nil {
+		return driver.WithStack(err)
+	}
+	return nil
+}
+
+// WriteKeyIfEmpty writes the given value with the given key only if the key was empty before.
+func (c *agency) WriteKeyIfEmpty(ctx context.Context, key []string, value interface{}, ttl time.Duration) error {
```

Comment:
Here, a retry can be disastrous, since the operation is not idempotent. This is actually a very difficult problem. There is the possibility to send a client ID along with a transaction, such that one is later able to enquire if an operation has actually happened already. However, this makes the client code quite difficult to get right (see our `AgencyComm`, we had considerable trouble to implement this correctly, LOL, and I am not sure we got it right now). Furthermore, even that is not entirely water-tight because of the rather fundamental two generals problem. We never know whether the first request is sent of and lingers in some TCP buffer, and is sent after a longish delay. During the time when it is in the buffers, it is not possible to either stop it or to find out if it will be executed eventually. I do not know what the right approach is here.

Reply:
See above